### PR TITLE
Restore Composer History under shift-up & down

### PR DESCRIPTION
**Pull request by t3chguy**
Fixes https://github.com/vector-im/riot-web/issues/9977

Shift-up is imperfect as spamming shift-up will not go up multiple history entries as it'll highlight a line in your history instead, maybe an alternative modifier would be better

Signed-off-by: Michael Telatynski <7t3chguy@gmail.com>

### CODE REVIEW

**Comment by bwindels** [comment]
So, after some discussion (on the internal room, sorry, should have taken riot-dev) we've decided we can steal the alt+up/down shortcut from moving between rooms, as that is unusable as it is anyway. You want to make that change, or shall I finish it off?

**Comment by t3chguy** [comment]
I will make it within a couple of hours :)

**Comment by bwindels** [comment]
> I will make it within a couple of hours :)

Sure, thanks!

File: src/components/views/rooms/MessageComposerInput.js [hunk 1]
```diff
@@ -1204,50 +1204,44 @@ export default class MessageComposerInput extends React.Component {
...
+                return;
             }
+        }
 
-            const selected = this.selectHistory(up);
-            if (selected) {
-                // We're selecting history, so prevent the key event from doing anything else
-                e.preventDefault();
-            }
+        const selected = this.selectHistory(up);
```
Comment (bwindels) [inline src/components/views/rooms/MessageComposerInput.js]: the logic gets a bit "double negative" and scattered in here, in that it's not immediately obvious that this code will only run when alt is pressed. How about:

```
if (e.altKey) {
   // restore history
} else if (up && !(e.ctrlKey || e.metaKey || e.shiftKey) && caretAtStart) {
  // edit last sent event
}
```

File: src/components/views/rooms/MessageComposerInput.js [hunk 2]
```diff
@@ -1206,15 +1206,22 @@ export default class MessageComposerInput extends React.Component {
     onVerticalArrow = (e, up) => {
         if (e.ctrlKey || e.shiftKey || e.metaKey) return;
 
-        // Select history
-        const selection = this.state.editorState.selection;
+        if (e.altKey) {
+            // Try select composer history
+            const selected = this.selectHistory(up);
```
Comment (bwindels) [inline src/components/views/rooms/MessageComposerInput.js]: shouldn't this also only happen if the selection is collapsed?

File: src/components/views/rooms/MessageComposerInput.js [hunk 3]
```diff
@@ -1206,15 +1206,22 @@ export default class MessageComposerInput extends React.Component {
...
+            // selection must be collapsed
+            if (!selection.isCollapsed) return;
+            const document = this.state.editorState.document;
 
-        // and we must be at the edge of the document (up=start, down=end)
-        if (up) {
+            // and we must be at the edge of the document (up=start, down=end)
             if (!selection.anchor.isAtStartOfNode(document)) return;
 
             if (!e.altKey) {
```
Comment (bwindels) [inline src/components/views/rooms/MessageComposerInput.js]: this is redundant now

**Comment by bwindels** [comment]
merged this, as I need to bring the history back as part of feature guarding up-to-edit ...

**Comment by t3chguy** [comment]
Sorry about not having time to bring it up to scratch. I'm future feel free to commit onto my branches to keep the changes consistent. Thanks for taking this on

**Comment by bwindels** [comment]
> Sorry about not having time to bring it up to scratch. I'm future feel free to commit onto my branches to keep the changes consistent. Thanks for taking this on

All good, will do :+1:

**Pull request merged**